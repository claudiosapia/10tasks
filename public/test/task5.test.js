const task5 = require("../task5");
//init max score to use in unit test
const maxScore = 200;
const students = task5.students;





//calc percentage of marks dividing  n. of correct answers by 200 wich is the maximum possible score multipling result by 100.
test("calcPercentageMark function", () => {
  const addMock = jest.spyOn(task5, "calcPercentageMark");
  // result will be  calcPercentageMarc(Math.floor(Math.random() * (maxScore - 60) + 60))
  const result = addMock(Math.floor(Math.random() * (maxScore - 60) + 60));
  const percentageMark=  Math.floor(result);
  expect(typeof percentageMark).toBe("number");
  expect(percentageMark).toBeGreaterThanOrEqual(1);
  expect(percentageMark).toBeLessThanOrEqual(100);
  expect(percentageMark).toMatchSnapshot();
});



//test randomMarks () that loops through students, generates random marks, convert in percentage using calcPercentageMark()

test("randomMarks function works as expected", () => {
  //spyOn method on task5, funct randomMarks
  const addMock = jest.spyOn(task5, "randomMarks");
  // result will be randomMarks(min) which is 60
  const result = addMock(60);
  //expect function to have been called
  expect(addMock).toHaveBeenCalled();
  //loop through function and check that percentage mark is a number
  for (let i = 0; i < students.length; i++) {
    expect(typeof students[i][2]).toBe("number");
    
          expect(typeof students[i][2]).toMatchSnapshot()
  }

});

test("returned value of marks inside marks object is correct", () => {
  //import marks
  const marks = task5.marks;

  //loop through object and check that...

  //value of marks within marks object
  for (const prop in marks) {
    //is a number
    expect(typeof marks[prop]).toBe("number");
    //is < or = 15 as there are 15 students there should be 15 marks
    expect(marks[prop]).toBeLessThanOrEqual(15);
    //is >= 0 as there are 15 students and there should be 15 marks
    expect(marks[prop]).toBeGreaterThanOrEqual(0);
    expect(marks[prop]).toMatchSnapshot()
  }
});


describe("students are sorted and displayed correctly", () => {


  it("checks that number of students is correct", () => {
    expect(students.length).toBe(15)
        expect(students.length).toMatchSnapshot()

      });


  
  it("should return sorted students by marks from high to low", () => {
    //loop through students to check if mark is >= 70%
    for (let i = 0; i < students.length; i++) {
      students.sort(function (a, b) {
        return b[2] - a[2];
      });
      //sort students by percentage from high to low
      let sortedStudents = students.sort(function (a, b) {
        return b[2] - a[2];
      });
        expect(sortedStudents).toMatchSnapshot()

      expect(sortedStudents[0][2]).toBeGreaterThanOrEqual(sortedStudents[1][2]);

      expect(sortedStudents[1][2]).toBeGreaterThanOrEqual(sortedStudents[2][2]);
      expect(sortedStudents[2][2]).toBeGreaterThanOrEqual(sortedStudents[3][2]);
      expect(sortedStudents[3][2]).toBeGreaterThanOrEqual(sortedStudents[4][2]);
      expect(sortedStudents[4][2]).toBeGreaterThanOrEqual(sortedStudents[5][2]);
    }
  });
});
