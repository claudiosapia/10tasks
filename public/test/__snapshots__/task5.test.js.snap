// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calcPercentageMark function 1`] = `75`;

exports[`randomMarks function works as expected 1`] = `"number"`;

exports[`randomMarks function works as expected 2`] = `"number"`;

exports[`randomMarks function works as expected 3`] = `"number"`;

exports[`randomMarks function works as expected 4`] = `"number"`;

exports[`randomMarks function works as expected 5`] = `"number"`;

exports[`randomMarks function works as expected 6`] = `"number"`;

exports[`randomMarks function works as expected 7`] = `"number"`;

exports[`randomMarks function works as expected 8`] = `"number"`;

exports[`randomMarks function works as expected 9`] = `"number"`;

exports[`randomMarks function works as expected 10`] = `"number"`;

exports[`randomMarks function works as expected 11`] = `"number"`;

exports[`randomMarks function works as expected 12`] = `"number"`;

exports[`randomMarks function works as expected 13`] = `"number"`;

exports[`randomMarks function works as expected 14`] = `"number"`;

exports[`randomMarks function works as expected 15`] = `"number"`;

exports[`returned value of marks inside marks object is correct 1`] = `6`;

exports[`returned value of marks inside marks object is correct 2`] = `1`;

exports[`returned value of marks inside marks object is correct 3`] = `3`;

exports[`returned value of marks inside marks object is correct 4`] = `3`;

exports[`returned value of marks inside marks object is correct 5`] = `2`;

exports[`students are sorted and displayed correctly checks that number of students is correct 1`] = `15`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 1`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 2`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 3`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 4`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 5`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 6`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 7`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 8`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 9`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 10`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 11`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 12`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 13`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 14`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;

exports[`students are sorted and displayed correctly should return sorted students by marks from high to low 15`] = `
Array [
  Array [
    "Elise",
    "Marks",
    98,
  ],
  Array [
    "Riyad",
    "Munro",
    96,
  ],
  Array [
    "Elsie",
    "Smith",
    94,
  ],
  Array [
    "Giovanni",
    "Armstrong",
    90,
  ],
  Array [
    "Hareem",
    "Bain",
    84,
  ],
  Array [
    "Jarod",
    "Bird",
    83,
  ],
  Array [
    "Philippe",
    "Ludacre",
    71,
  ],
  Array [
    "Shaquille",
    "Hodgson",
    70,
  ],
  Array [
    "Hazel",
    "Wainwright,",
    62,
  ],
  Array [
    "Tymon",
    "Jeffery",
    56,
  ],
  Array [
    "Ethel",
    "Hickman",
    53,
  ],
  Array [
    "Sebastian",
    "Papale",
    41,
  ],
  Array [
    "Cristian",
    "Larson",
    39,
  ],
  Array [
    "Nick",
    "Muller",
    37,
  ],
  Array [
    "Declan",
    "Lucero",
    32,
  ],
]
`;
